subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ZipHasher.java

```diff
@@ -32,82 +34,98 @@
 
-    private final ResourceHasher classpathResourceHasher;
-    private final ResourceFilter classpathResourceFilter;
+    private static final Set<String> KNOWN_ZIP_EXTENSIONS = ImmutableSet.of(".zip", ".jar", ".war", ".ear", ".apk");
+    private static final Logger LOGGER = LoggerFactory.getLogger(ZipHasher.class);
 
-    public JarHasher(ResourceHasher classpathResourceHasher, ResourceFilter classpathResourceFilter) {
-        this.classpathResourceHasher = classpathResourceHasher;
-        this.classpathResourceFilter = classpathResourceFilter;
+    public static boolean isZipFile(final String name) {
+        for (String extension : KNOWN_ZIP_EXTENSIONS) {
+            if (FileUtils.hasExtensionIgnoresCase(name, extension)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private final ResourceHasher resourceHasher;
+    private final ResourceFilter resourceFilter;
+
+    public ZipHasher(ResourceHasher resourceHasher, ResourceFilter resourceFilter) {
+        this.resourceHasher = resourceHasher;
+        this.resourceFilter = resourceFilter;
     }
 
     @Nullable
     @Override
     public HashCode hash(RegularFileSnapshot fileSnapshot) {
-        return hashJarContents(fileSnapshot);
+        return hashZipContents(fileSnapshot);
     }
 
     @Override
     public void appendConfigurationToHasher(Hasher hasher) {
         hasher.putString(getClass().getName());
-        classpathResourceHasher.appendConfigurationToHasher(hasher);
-        classpathResourceFilter.appendConfigurationToHasher(hasher);
+        resourceHasher.appendConfigurationToHasher(hasher);
+        resourceFilter.appendConfigurationToHasher(hasher);
     }
 
     @Nullable
-    private HashCode hashJarContents(RegularFileSnapshot jarFileSnapshot) {
+    private HashCode hashZipContents(RegularFileSnapshot zipFileSnapshot) {
         try {
-            List<FileSystemLocationFingerprint> fingerprints = fingerprintZipEntries(jarFileSnapshot.getAbsolutePath());
+            List<FileSystemLocationFingerprint> fingerprints = fingerprintZipEntries(zipFileSnapshot.getAbsolutePath());
             if (fingerprints.isEmpty()) {
                 return null;
             }
             Hasher hasher = Hashing.newHasher();
             NormalizedPathFingerprintCompareStrategy.appendSortedToHasher(hasher, fingerprints);
             return hasher.hash();
         } catch (Exception e) {
-            return hashMalformedZip(jarFileSnapshot, e);
+            return hashMalformedZip(zipFileSnapshot, e);
         }
     }
 
-    private List<FileSystemLocationFingerprint> fingerprintZipEntries(String jarFile) throws IOException {
-        List<FileSystemLocationFingerprint> fingerprints = Lists.newArrayList();
-        InputStream fileInputStream = null;
+    private List<FileSystemLocationFingerprint> fingerprintZipEntries(String zipFile) throws IOException {
+        ZipInput input = null;
```

Comment:
Probably not for this PR, but we should move this to `:snapshots` and then we could do:

```java
try (ZipInput input = new FileZipInput(new File(zipFile))) {
    return fingerprintZipEntries(input);
}
```